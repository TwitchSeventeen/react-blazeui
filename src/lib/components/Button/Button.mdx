---
name: Button
menu: Components
---

import { Playground, PropsTable } from 'docz';
import Button from './index';

# Button

Probably the first thing people look at when choosing a UI toolkit. The common button!

## Button Colours 
Use the type prop to provide styling to the alert.
<Playground>
    <Button>Default</Button>
    <br /><br />
    <Button type="brand">Brand</Button>
    <br /><br />
    <Button type="info">Brand</Button>
    <br /><br />
    <Button type="warning">Warning</Button>
    <br /><br />
    <Button type="success">Success</Button>
    <br /><br />
    <Button type="error">Error</Button>
</Playground>

## Button Groups
To group related buttons together you can use the Button.Group element. This will bunch the buttons together and remove the spacing between them.
<Playground>
    <Button.Group>
        <Button type="brand">Button</Button>
        <Button>Button</Button>
    </Button.Group>
</Playground>

Adding the **rounded** prop will make the first and last buttons rounded.
<Playground>
    <Button.Group rounded>
        <Button type="brand">Button</Button>
        <Button>Button</Button>
        <Button type="info">Button</Button>
    </Button.Group>
</Playground>

## Button Sizes 
Adding a **size** utility prop will change the size of the button.
<Playground>
    <Button size="xsmall">xsmall</Button>
    <Button size="small">small</Button>
    <Button size="medium">medium</Button>
    <Button size="large">large</Button>
    <Button size="xlarge">xlarge</Button>
    <Button size="super">super</Button>
</Playground>

## Full Width Buttons  
<Playground>
    <Button size="xsmall" fluid>xsmall</Button>
    <Button size="small" fluid>small</Button>
    <Button size="medium" fluid>medium</Button>
    <Button size="large" fluid>large</Button>
    <Button size="xlarge" fluid>xlarge</Button>
    <Button size="super" fluid>super</Button>
</Playground>

## Ghost Buttons
Use the type prop to provide styling to the alert.
<Playground>
    <Button size="xsmall" ghost>xsmall</Button>
    <br /><br />
    <Button size="small" ghost>small</Button>
    <br /><br />
    <Button size="medium" ghost>medium</Button>
    <br /><br />
    <Button size="large" ghost>large</Button>
    <br /><br />
    <Button size="xlarge" ghost>xlarge</Button>
    <br /><br />
    <Button size="super" ghost>super</Button>
</Playground>

## Ghost Colours
<Playground>
    <Button ghost>Default</Button>
    <br /><br />
    <Button type="brand" ghost>Brand</Button>
    <br /><br />
    <Button type="info" ghost>Brand</Button>
    <br /><br />
    <Button type="warning" ghost>Warning</Button>
    <br /><br />
    <Button type="success" ghost>Success</Button>
    <br /><br />
    <Button type="error" ghost>Error</Button>
</Playground>

## Rounded Buttons
<Playground>
    <Button ghost rounded>Default</Button>
    <br /><br />
    <Button type="brand" ghost rounded>Brand</Button>
    <br /><br />
    <Button type="info" ghost rounded>Brand</Button>
    <br /><br />
    <Button type="warning" ghost rounded>Warning</Button>
    <br /><br />
    <Button type="success" ghost rounded>Success</Button>
    <br /><br />
    <Button type="error" ghost rounded>Error</Button>
</Playground>

## Active state
<Playground>
    <Button.Group>
        <Button type="info" active>Button 1</Button>
        <Button type="info">Button 2</Button>
    </Button.Group>
</Playground>

## Button Type
By default buttons are rendered with a HTML type of _button_. Use the **buttonType** prop to change this to _submit_ or _reset_.

## Properties
### Button
<PropsTable of={Button} />

### Button.Group
<PropsTable of={Button.Group} />