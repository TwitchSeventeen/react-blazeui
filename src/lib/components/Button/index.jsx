import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Group from '../Input/Group';
import omitOwnProps from '../../utils/omitOwnProps';

function Button(props) {
  const {
    children, className, buttonType, type, size, fluid, ghost, rounded, active,
  } = props;
  const classes = classNames(
    'c-button',
    { [`c-button--${type}`]: (type && !ghost) },
    { 'c-button--ghost': (!type && ghost) },
    { [`c-button--ghost-${type}`]: (type && ghost) },
    { 'c-button--block': fluid },
    { 'c-button--rounded': rounded },
    { 'c-button--active': active },
    { [`u-${size}`]: size },
    className,
  );
  const spreadProps = omitOwnProps(props);
  return (
    <button {...spreadProps} type={buttonType} className={classes}>
      { children }
    </button>
  );
}

Button.propTypes = {
  /**
   * This element can have children
   */
  children: PropTypes.node,
  /**
   * Additional classes
   */
  className: PropTypes.string,
  /**
   * The type of HTML button
   */
  buttonType: PropTypes.oneOf([
    'button',
    'submit',
    'reset',
  ]),
  /**
   * The type of button
   */
  type: PropTypes.oneOf([
    'brand',
    'info',
    'warning',
    'success',
    'error',
  ]),
  /**
   * The type of button
   */
  size: PropTypes.oneOf([
    'xsmall',
    'small',
    'medium',
    'large',
    'xlarge',
    'super',
  ]),
  /**
   * Defines a full width button
   */
  fluid: PropTypes.bool,
  /**
   * Defines a rounded button
   */
  rounded: PropTypes.bool,
  /**
   * Defines an active state
   */
  active: PropTypes.bool,
  /**
   * Defines an ghost state
   */
  ghost: PropTypes.bool,
};

Button.defaultProps = {
  children: null,
  buttonType: 'button',
  type: null,
  size: null,
  fluid: false,
  rounded: false,
  active: false,
  className: null,
  ghost: false,
};

Button.Group = Group;

export default Button;
