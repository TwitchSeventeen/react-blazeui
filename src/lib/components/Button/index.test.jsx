import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Button from './index';

describe('Button', () => {
  it('should be defined', () => {
    expect(Button).toBeDefined();
  });

  it('renders correctly', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <Button>
          lorem ipsum dolor
      </Button>,
    );

    const tree = renderer.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly a ghost button', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <Button ghost>
          lorem ipsum dolor
      </Button>,
    );

    const tree = renderer.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly a ghost button with a type', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <Button ghost type="success">
          lorem ipsum dolor
      </Button>,
    );

    const tree = renderer.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly a button with a type', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <Button type="success">
          lorem ipsum dolor
      </Button>,
    );

    const tree = renderer.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });
});
