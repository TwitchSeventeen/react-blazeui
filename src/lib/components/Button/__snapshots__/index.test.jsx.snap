// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<button
  active={false}
  buttonType="button"
  className="c-button"
  fluid={false}
  ghost={false}
  rounded={false}
  size={null}
  type="button"
>
  lorem ipsum dolor
</button>
`;

exports[`Button renders correctly a button with a type 1`] = `
<button
  active={false}
  buttonType="button"
  className="c-button c-button--success"
  fluid={false}
  ghost={false}
  rounded={false}
  size={null}
  type="button"
>
  lorem ipsum dolor
</button>
`;

exports[`Button renders correctly a ghost button 1`] = `
<button
  active={false}
  buttonType="button"
  className="c-button c-button--ghost"
  fluid={false}
  ghost={true}
  rounded={false}
  size={null}
  type="button"
>
  lorem ipsum dolor
</button>
`;

exports[`Button renders correctly a ghost button with a type 1`] = `
<button
  active={false}
  buttonType="button"
  className="c-button c-button--ghost-success"
  fluid={false}
  ghost={true}
  rounded={false}
  size={null}
  type="button"
>
  lorem ipsum dolor
</button>
`;
